File: java/src/main/java/io/github/nickliapin/tdc/unicode/Alphabets.java
package io.github.nickliapin.tdc.unicode;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The named alphabets a config can ask for by name.
 *
 * <p>Spelled out as explicit code-point ranges rather than looked up through Unicode character
 * properties. Property tables ship with the runtime and change between versions of it, so a
 * config that drew Cyrillic letters could quietly draw a different set of them after an upgrade,
 * and two languages' runtimes would never agree. A range written here means the same thing
 * everywhere, forever.
 */
public final class Alphabets {

  private static final Map<String, List<String>> REGISTRY = new LinkedHashMap<>();

  static {
    List<String> latinLower = between('a', 'z');
    List<String> latinUpper = between('A', 'Z');
    List<String> digits = between('0', '9');

    // Ё sits outside the alphabetical block in Unicode but inside it in the alphabet, so it is
    // spliced back into place rather than appended.
    List<String> cyrLower = new ArrayList<>(between('а', 'е'));
    cyrLower.add("ё");
    cyrLower.addAll(between('ж', 'я'));
    List<String> cyrUpper = new ArrayList<>(between('А', 'Е'));
    cyrUpper.add("Ё");
    cyrUpper.addAll(between('Ж', 'Я'));

    REGISTRY.put("latin.lower", latinLower);
    REGISTRY.put("latin.upper", latinUpper);
    REGISTRY.put("latin.letters", concat(latinUpper, latinLower));
    REGISTRY.put("digits.ascii", digits);
    REGISTRY.put("digits.fullwidth", between('０', '９'));
    REGISTRY.put("cyrillic.ru.lower", List.copyOf(cyrLower));
    REGISTRY.put("cyrillic.ru.upper", List.copyOf(cyrUpper));
    REGISTRY.put("cyrillic.ru.letters", concat(cyrUpper, cyrLower));
    REGISTRY.put("greek.letters", codePoints("ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩαβγδεζηθικλμνξοπρσςτυφχψω"));
    REGISTRY.put("hebrew.letters", between('א', 'ת'));
    REGISTRY.put("arabic.letters", between('ء', 'ي'));
    REGISTRY.put("kana.hiragana", between('ぁ', 'ゖ'));
    REGISTRY.put("kana.katakana", between('ァ', 'ヺ'));
    REGISTRY.put("cjk.unified.basic", between('一', '鿿'));
    REGISTRY.put("roman.upper", codePoints("IVXLCDM"));
    REGISTRY.put("roman.lower", codePoints("ivxlcdm"));
  }

  private Alphabets() {}

  /** {@code null} when the name is unknown; callers report it with the list of known names. */
  public static List<String> chars(String name) {
    return REGISTRY.get(name);
  }

  public static List<String> names() {
    return List.copyOf(REGISTRY.keySet());
  }

  public static List<String> between(int start, int end) {
    if (start > end) {
      throw new IllegalArgumentException("invalid alphabet range");
    }
    List<String> out = new ArrayList<>(end - start + 1);
    for (int cp = start; cp <= end; cp++) {
      out.add(new String(Character.toChars(cp)));
    }
    return out;
  }

  /** One entry per code point, so characters outside the basic plane stay whole. */
  public static List<String> codePoints(String value) {
    List<String> out = new ArrayList<>();
    value.codePoints().forEach(cp -> out.add(new String(Character.toChars(cp))));
    return out;
  }

  private static List<String> concat(List<String> a, List<String> b) {
    List<String> out = new ArrayList<>(a.size() + b.size());
    out.addAll(a);
    out.addAll(b);
    return List.copyOf(out);
  }
}

File: typescript/src/unicode/alphabets.ts
/**
 * Named Unicode alphabets shared by character-oriented generators.
 *
 * Keep this registry intentionally explicit. It is easier to port to
 * Python/Java than relying on runtime-specific Unicode property engines,
 * and it gives the DSL stable names that can be documented and tested.
 */

export interface AlphabetDefinition {
  readonly name: string;
  readonly description: string;
  readonly chars: readonly string[];
}

const ASCII_DIGITS = charsBetween('0', '9');
const LATIN_LOWER = charsBetween('a', 'z');
const LATIN_UPPER = charsBetween('A', 'Z');

const CYRILLIC_RU_LOWER = [...charsBetween('а', 'е'), 'ё', ...charsBetween('ж', 'я')] as const;
const CYRILLIC_RU_UPPER = [...charsBetween('А', 'Е'), 'Ё', ...charsBetween('Ж', 'Я')] as const;

const DEFINITIONS: readonly AlphabetDefinition[] = [
  {
    name: 'latin.lower',
    description: 'ASCII lowercase Latin letters a-z',
    chars: LATIN_LOWER,
  },
  {
    name: 'latin.upper',
    description: 'ASCII uppercase Latin letters A-Z',
    chars: LATIN_UPPER,
  },
  {
    name: 'latin.letters',
    description: 'ASCII Latin letters A-Z and a-z',
    chars: [...LATIN_UPPER, ...LATIN_LOWER],
  },
  {
    name: 'digits.ascii',
    description: 'ASCII digits 0-9',
    chars: ASCII_DIGITS,
  },
  {
    name: 'digits.fullwidth',
    description: 'Fullwidth digits ０-９',
    chars: charsBetween('０', '９'),
  },
  {
    name: 'cyrillic.ru.lower',
    description: 'Russian lowercase Cyrillic letters, including ё',
    chars: CYRILLIC_RU_LOWER,
  },
  {
    name: 'cyrillic.ru.upper',
    description: 'Russian uppercase Cyrillic letters, including Ё',
    chars: CYRILLIC_RU_UPPER,
  },
  {
    name: 'cyrillic.ru.letters',
    description: 'Russian Cyrillic letters, including Ё/ё',
    chars: [...CYRILLIC_RU_UPPER, ...CYRILLIC_RU_LOWER],
  },
  {
    name: 'greek.letters',
    description: 'Basic Greek letters Α-Ω and α-ω',
    chars: Array.from('ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩαβγδεζηθικλμνξοπρσςτυφχψω'),
  },
  {
    name: 'hebrew.letters',
    description: 'Hebrew letters א-ת',
    chars: charsBetween('א', 'ת'),
  },
  {
    name: 'arabic.letters',
    description: 'Arabic letters ء-ي',
    chars: charsBetween('ء', 'ي'),
  },
  {
    name: 'kana.hiragana',
    description: 'Japanese Hiragana characters ぁ-ゖ',
    chars: charsBetween('ぁ', 'ゖ'),
  },
  {
    name: 'kana.katakana',
    description: 'Japanese Katakana characters ァ-ヺ',
    chars: charsBetween('ァ', 'ヺ'),
  },
  {
    name: 'cjk.unified.basic',
    description: 'CJK Unified Ideographs U+4E00..U+9FFF',
    chars: charsBetween('\u4E00', '\u9FFF'),
  },
  {
    name: 'roman.upper',
    description: 'Roman numeral letters I, V, X, L, C, D, M',
    chars: Array.from('IVXLCDM'),
  },
  {
    name: 'roman.lower',
    description: 'Roman numeral letters i, v, x, l, c, d, m',
    chars: Array.from('ivxlcdm'),
  },
] as const;

const REGISTRY = new Map<string, AlphabetDefinition>(
  DEFINITIONS.map((definition) => [definition.name, definition]),
);

export const ALPHABET_DEFINITIONS: readonly AlphabetDefinition[] = DEFINITIONS;
export const ALPHABET_NAMES: readonly string[] = DEFINITIONS.map((definition) => definition.name);

export function resolveAlphabet(name: string): AlphabetDefinition | undefined {
  return REGISTRY.get(name);
}

export function resolveAlphabetChars(name: string): readonly string[] | undefined {
  return resolveAlphabet(name)?.chars;
}

export function charsBetween(start: string, end: string): string[] {
  const a = start.codePointAt(0);
  const b = end.codePointAt(0);
  if (a === undefined || b === undefined || a > b) {
    throw new Error(`invalid alphabet range "${start}-${end}"`);
  }
  const out: string[] = [];
  for (let code = a; code <= b; code++) out.push(String.fromCodePoint(code));
  return out;
}

export function toCodePoints(value: string): string[] {
  return Array.from(value);
}
